core/vm/runtime/runtime.go

```diff
@@ -101,6 +101,10 @@ func setDefaults(cfg *Config) {
 	if cfg.BlobBaseFee == nil {
 		cfg.BlobBaseFee = big.NewInt(params.BlobTxMinBlobGasprice)
 	}
+	// Merge indicators
+	if t := cfg.ChainConfig.ShanghaiTime; cfg.ChainConfig.TerminalTotalDifficultyPassed || (t != nil && *t == 0) {
```

Comment:
It seems like block number can be overriden. Is the *t == 0 check enough?

Reply:
Shouldn't we also enable these forks by default?

e.g. 

```diff
diff --git a/core/vm/runtime/runtime.go b/core/vm/runtime/runtime.go
index b587d6d5a0..9c448a80c7 100644
--- a/core/vm/runtime/runtime.go
+++ b/core/vm/runtime/runtime.go
@@ -57,21 +57,32 @@ type Config struct {
 // sets defaults on the config
 func setDefaults(cfg *Config) {
 	if cfg.ChainConfig == nil {
+		var (
+			shanghaiTime = uint64(0)
+			cancunTime   = uint64(0)
+		)
 		cfg.ChainConfig = &params.ChainConfig{
-			ChainID:             big.NewInt(1),
-			HomesteadBlock:      new(big.Int),
-			DAOForkBlock:        new(big.Int),
-			DAOForkSupport:      false,
-			EIP150Block:         new(big.Int),
-			EIP155Block:         new(big.Int),
-			EIP158Block:         new(big.Int),
-			ByzantiumBlock:      new(big.Int),
-			ConstantinopleBlock: new(big.Int),
-			PetersburgBlock:     new(big.Int),
-			IstanbulBlock:       new(big.Int),
-			MuirGlacierBlock:    new(big.Int),
-			BerlinBlock:         new(big.Int),
-			LondonBlock:         new(big.Int),
+			ChainID:                       big.NewInt(1),
+			HomesteadBlock:                new(big.Int),
+			DAOForkBlock:                  new(big.Int),
+			DAOForkSupport:                false,
+			EIP150Block:                   new(big.Int),
+			EIP155Block:                   new(big.Int),
+			EIP158Block:                   new(big.Int),
+			ByzantiumBlock:                new(big.Int),
+			ConstantinopleBlock:           new(big.Int),
+			PetersburgBlock:               new(big.Int),
+			IstanbulBlock:                 new(big.Int),
+			MuirGlacierBlock:              new(big.Int),
+			BerlinBlock:                   new(big.Int),
+			LondonBlock:                   new(big.Int),
+			ArrowGlacierBlock:             nil,
+			GrayGlacierBlock:              nil,
+			TerminalTotalDifficulty:       big.NewInt(0),
+			TerminalTotalDifficultyPassed: true,
+			MergeNetsplitBlock:            nil,
+			ShanghaiTime:                  &shanghaiTime,
+			CancunTime:                    &cancunTime,
 		}
 	}
 
diff --git a/triedb/database.go b/triedb/database.go
index 10f77982f3..b1643de3ed 100644
--- a/triedb/database.go
+++ b/triedb/database.go
@@ -266,14 +266,7 @@ func (db *Database) Recover(target common.Hash) error {
 	if !ok {
 		return errors.New("not supported")
 	}
-	var loader triestate.TrieLoader
-	if db.config.IsVerkle {
-		// TODO define verkle loader
-		log.Crit("Verkle loader is not defined")
-	} else {
-		loader = trie.NewMerkleLoader(db)
-	}
-	return pdb.Recover(target, loader)
+	return pdb.Recover(db, target)
 }
 
 // Recoverable returns the indicator if the specified state is enabled to be
diff --git a/triedb/pathdb/database.go b/triedb/pathdb/database.go
index 05a28aa1ef..bf100cdd6b 100644
--- a/triedb/pathdb/database.go
+++ b/triedb/pathdb/database.go
@@ -19,6 +19,7 @@ package pathdb
 import (
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/triedb/database"
 	"io"
 	"sync"
 	"time"
@@ -345,7 +346,7 @@ func (db *Database) Enable(root common.Hash) error {
 // Recover rollbacks the database to a specified historical point.
 // The state is supported as the rollback destination only if it's
 // canonical state and the corresponding trie histories are existent.
-func (db *Database) Recover(root common.Hash, loader triestate.TrieLoader) error {
+func (db *Database) Recover(d database.Database, root common.Hash) error {
 	db.lock.Lock()
 	defer db.lock.Unlock()
 
@@ -371,7 +372,7 @@ func (db *Database) Recover(root common.Hash, loader triestate.TrieLoader) error
 		if err != nil {
 			return err
 		}
-		dl, err = dl.revert(h, loader)
+		dl, err = dl.revert(d, h)
 		if err != nil {
 			return err
 		}
diff --git a/triedb/pathdb/disklayer.go b/triedb/pathdb/disklayer.go
index 964ad2ef77..4efca12222 100644
--- a/triedb/pathdb/disklayer.go
+++ b/triedb/pathdb/disklayer.go
@@ -18,6 +18,7 @@ package pathdb
 
 import (
 	"fmt"
+	"github.com/ethereum/go-ethereum/triedb/database"
 	"sync"
 
 	"github.com/VictoriaMetrics/fastcache"
@@ -219,7 +220,7 @@ func (dl *diskLayer) commit(bottom *diffLayer, force bool) (*diskLayer, error) {
 }
 
 // revert applies the given state history and return a reverted disk layer.
-func (dl *diskLayer) revert(h *history, loader triestate.TrieLoader) (*diskLayer, error) {
+func (dl *diskLayer) revert(db database.Database, h *history) (*diskLayer, error) {
 	if h.meta.root != dl.rootHash() {
 		return nil, errUnexpectedHistory
 	}
@@ -229,7 +230,7 @@ func (dl *diskLayer) revert(h *history, loader triestate.TrieLoader) (*diskLayer
 	// Apply the reverse state changes upon the current state. This must
 	// be done before holding the lock in order to access state in "this"
 	// layer.
-	nodes, err := triestate.Apply(h.meta.parent, h.meta.root, h.accounts, h.storages, loader)
+	nodes, err := apply(db, h.meta.parent, h.meta.root, h.accounts, h.storages)
 	if err != nil {
 		return nil, err
 	}
diff --git a/triedb/pathdb/state_exec.go b/triedb/pathdb/state_exec.go
index 4b31220ed6..69b0a90dca 100644
--- a/triedb/pathdb/state_exec.go
+++ b/triedb/pathdb/state_exec.go
@@ -1 +1,218 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
 package pathdb
+
+import (
+	"errors"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-ethereum/triedb/database"
+)
+
+// Trie is an Ethereum state trie, can be implemented by Ethereum Merkle Patricia
+// tree or Verkle tree.
+type Trie interface {
+	// Get returns the value for key stored in the trie.
+	Get(key []byte) ([]byte, error)
+
+	// Update associates key with value in the trie.
+	Update(key, value []byte) error
+
+	// Delete removes any existing value for key from the trie.
+	Delete(key []byte) error
+
+	// Commit the trie and returns a set of dirty nodes generated along with
+	// the new root hash.
+	Commit(collectLeaf bool) (common.Hash, *trienode.NodeSet, error)
+}
+
+// TrieLoader wraps functions to load tries.
+type TrieLoader interface {
+	// OpenTrie opens the main account trie.
+	OpenTrie(root common.Hash) (Trie, error)
+
+	// OpenStorageTrie opens the storage trie of an account.
+	OpenStorageTrie(stateRoot common.Hash, addrHash, root common.Hash) (Trie, error)
+}
+
+// context wraps all fields for executing state diffs.
+type context struct {
+	prevRoot    common.Hash
+	postRoot    common.Hash
+	accounts    map[common.Address][]byte
+	storages    map[common.Address]map[common.Hash][]byte
+	accountTrie Trie
+	nodes       *trienode.MergedNodeSet
+}
+
+// apply traverses the provided state diffs, apply them in the associated
+// post-state and return the generated dirty trie nodes. The state can be
+// loaded via the provided trie loader.
+func apply(db database.Database, prevRoot common.Hash, postRoot common.Hash, accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte) (map[common.Hash]map[string]*trienode.Node, error) {
+	tr, err := trie.New(trie.TrieID(postRoot), db)
+	if err != nil {
+		return nil, err
+	}
+	ctx := &context{
+		prevRoot:    prevRoot,
+		postRoot:    postRoot,
+		accounts:    accounts,
+		storages:    storages,
+		accountTrie: tr,
+		nodes:       trienode.NewMergedNodeSet(),
+	}
+	for addr, account := range accounts {
+		var err error
+		if len(account) == 0 {
+			err = deleteAccount(ctx, db, addr)
+		} else {
+			err = updateAccount(ctx, db, addr)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to revert state, err: %w", err)
+		}
+	}
+	root, result, err := tr.Commit(false)
+	if err != nil {
+		return nil, err
+	}
+	if root != prevRoot {
+		return nil, fmt.Errorf("failed to revert state, want %#x, got %#x", prevRoot, root)
+	}
+	if err := ctx.nodes.Merge(result); err != nil {
+		return nil, err
+	}
+	return ctx.nodes.Flatten(), nil
+}
+
+// updateAccount the account was present in prev-state, and may or may not
+// existent in post-state. Apply the reverse diff and verify if the storage
+// root matches the one in prev-state account.
+func updateAccount(ctx *context, db database.Database, addr common.Address) error {
+	// The account was present in prev-state, decode it from the
+	// 'slim-rlp' format bytes.
+	h := newHasher()
+	defer h.release()
+
+	addrHash := h.hash(addr.Bytes())
+	prev, err := types.FullAccount(ctx.accounts[addr])
+	if err != nil {
+		return err
+	}
+	// The account may or may not existent in post-state, try to
+	// load it and decode if it's found.
+	blob, err := ctx.accountTrie.Get(addrHash.Bytes())
+	if err != nil {
+		return err
+	}
+	post := types.NewEmptyStateAccount()
+	if len(blob) != 0 {
+		if err := rlp.DecodeBytes(blob, &post); err != nil {
+			return err
+		}
+	}
+	// Apply all storage changes into the post-state storage trie.
+	st, err := trie.New(trie.StorageTrieID(ctx.postRoot, addrHash, post.Root), db)
+	if err != nil {
+		return err
+	}
+	for key, val := range ctx.storages[addr] {
+		var err error
+		if len(val) == 0 {
+			err = st.Delete(key.Bytes())
+		} else {
+			err = st.Update(key.Bytes(), val)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	root, result, err := st.Commit(false)
+	if err != nil {
+		return err
+	}
+	if root != prev.Root {
+		return errors.New("failed to reset storage trie")
+	}
+	// The returned set can be nil if storage trie is not changed
+	// at all.
+	if result != nil {
+		if err := ctx.nodes.Merge(result); err != nil {
+			return err
+		}
+	}
+	// Write the prev-state account into the main trie
+	full, err := rlp.EncodeToBytes(prev)
+	if err != nil {
+		return err
+	}
+	return ctx.accountTrie.Update(addrHash.Bytes(), full)
+}
+
+// deleteAccount the account was not present in prev-state, and is expected
+// to be existent in post-state. Apply the reverse diff and verify if the
+// account and storage is wiped out correctly.
+func deleteAccount(ctx *context, db database.Database, addr common.Address) error {
+	// The account must be existent in post-state, load the account.
+	h := newHasher()
+	defer h.release()
+
+	addrHash := h.hash(addr.Bytes())
+	blob, err := ctx.accountTrie.Get(addrHash.Bytes())
+	if err != nil {
+		return err
+	}
+	if len(blob) == 0 {
+		return fmt.Errorf("account is non-existent %#x", addrHash)
+	}
+	var post types.StateAccount
+	if err := rlp.DecodeBytes(blob, &post); err != nil {
+		return err
+	}
+	st, err := trie.New(trie.StorageTrieID(ctx.postRoot, addrHash, post.Root), db)
+	if err != nil {
+		return err
+	}
+	for key, val := range ctx.storages[addr] {
+		if len(val) != 0 {
+			return errors.New("expect storage deletion")
+		}
+		if err := st.Delete(key.Bytes()); err != nil {
+			return err
+		}
+	}
+	root, result, err := st.Commit(false)
+	if err != nil {
+		return err
+	}
+	if root != types.EmptyRootHash {
+		return errors.New("failed to clear storage trie")
+	}
+	// The returned set can be nil if storage trie is not changed
+	// at all.
+	if result != nil {
+		if err := ctx.nodes.Merge(result); err != nil {
+			return err
+		}
+	}
+	// Delete the post-state account from the main trie.
+	return ctx.accountTrie.Delete(addrHash.Bytes())
+}
diff --git a/triedb/state.go b/triedb/state.go
index ef8676922f..f5d037a0bb 100644
--- a/triedb/state.go
+++ b/triedb/state.go
@@ -1 +1,37 @@
 package triedb
+
+import "github.com/ethereum/go-ethereum/common"
+
+// Set represents a collection of mutated states during a state transition.
+// The value refers to the original content of state before the transition
+// is made. Nil means that the state was not present previously.
+type Set struct {
+	Accounts map[common.Address][]byte                 // Mutated account set, nil means the account was not present
+	Storages map[common.Address]map[common.Hash][]byte // Mutated storage set, nil means the slot was not present
+	size     common.StorageSize                        // Approximate size of set
+}
+
+// New constructs the state set with provided data.
+func New(accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte) *Set {
+	return &Set{
+		Accounts: accounts,
+		Storages: storages,
+	}
+}
+
+// Size returns the approximate memory size occupied by the set.
+func (s *Set) Size() common.StorageSize {
+	if s.size != 0 {
+		return s.size
+	}
+	for _, account := range s.Accounts {
+		s.size += common.StorageSize(common.AddressLength + len(account))
+	}
+	for _, slots := range s.Storages {
+		for _, val := range slots {
+			s.size += common.StorageSize(common.HashLength + len(val))
+		}
+		s.size += common.StorageSize(common.AddressLength)
+	}
+	return s.size
+}
diff --git a/triedb/state/state.go b/triedb/state/state.go
index 7bf2df5b48..4e0bad5d7a 100644
--- a/triedb/state/state.go
+++ b/triedb/state/state.go
@@ -1 +1,238 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
 package state
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+)
+
+// Trie is an Ethereum state trie, can be implemented by Ethereum Merkle Patricia
+// tree or Verkle tree.
+type Trie interface {
+	// Get returns the value for key stored in the trie.
+	Get(key []byte) ([]byte, error)
+
+	// Update associates key with value in the trie.
+	Update(key, value []byte) error
+
+	// Delete removes any existing value for key from the trie.
+	Delete(key []byte) error
+
+	// Commit the trie and returns a set of dirty nodes generated along with
+	// the new root hash.
+	Commit(collectLeaf bool) (common.Hash, *trienode.NodeSet, error)
+}
+
+// TrieLoader wraps functions to load tries.
+type TrieLoader interface {
+	// OpenTrie opens the main account trie.
+	OpenTrie(root common.Hash) (Trie, error)
+
+	// OpenStorageTrie opens the storage trie of an account.
+	OpenStorageTrie(stateRoot common.Hash, addrHash, root common.Hash) (Trie, error)
+}
+
+// context wraps all fields for executing state diffs.
+type context struct {
+	prevRoot    common.Hash
+	postRoot    common.Hash
+	accounts    map[common.Address][]byte
+	storages    map[common.Address]map[common.Hash][]byte
+	accountTrie Trie
+	nodes       *trienode.MergedNodeSet
+}
+
+// Apply traverses the provided state diffs, apply them in the associated
+// post-state and return the generated dirty trie nodes. The state can be
+// loaded via the provided trie loader.
+func Apply(prevRoot common.Hash, postRoot common.Hash, accounts map[common.Address][]byte, storages map[common.Address]map[common.Hash][]byte, loader TrieLoader) (map[common.Hash]map[string]*trienode.Node, error) {
+	tr, err := loader.OpenTrie(postRoot)
+	if err != nil {
+		return nil, err
+	}
+	ctx := &context{
+		prevRoot:    prevRoot,
+		postRoot:    postRoot,
+		accounts:    accounts,
+		storages:    storages,
+		accountTrie: tr,
+		nodes:       trienode.NewMergedNodeSet(),
+	}
+	for addr, account := range accounts {
+		var err error
+		if len(account) == 0 {
+			err = deleteAccount(ctx, loader, addr)
+		} else {
+			err = updateAccount(ctx, loader, addr)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to revert state, err: %w", err)
+		}
+	}
+	root, result, err := tr.Commit(false)
+	if err != nil {
+		return nil, err
+	}
+	if root != prevRoot {
+		return nil, fmt.Errorf("failed to revert state, want %#x, got %#x", prevRoot, root)
+	}
+	if err := ctx.nodes.Merge(result); err != nil {
+		return nil, err
+	}
+	return ctx.nodes.Flatten(), nil
+}
+
+// updateAccount the account was present in prev-state, and may or may not
+// existent in post-state. Apply the reverse diff and verify if the storage
+// root matches the one in prev-state account.
+func updateAccount(ctx *context, loader TrieLoader, addr common.Address) error {
+	// The account was present in prev-state, decode it from the
+	// 'slim-rlp' format bytes.
+	h := newHasher()
+	defer h.release()
+
+	addrHash := h.hash(addr.Bytes())
+	prev, err := types.FullAccount(ctx.accounts[addr])
+	if err != nil {
+		return err
+	}
+	// The account may or may not existent in post-state, try to
+	// load it and decode if it's found.
+	blob, err := ctx.accountTrie.Get(addrHash.Bytes())
+	if err != nil {
+		return err
+	}
+	post := types.NewEmptyStateAccount()
+	if len(blob) != 0 {
+		if err := rlp.DecodeBytes(blob, &post); err != nil {
+			return err
+		}
+	}
+	// Apply all storage changes into the post-state storage trie.
+	st, err := loader.OpenStorageTrie(ctx.postRoot, addrHash, post.Root)
+	if err != nil {
+		return err
+	}
+	for key, val := range ctx.storages[addr] {
+		var err error
+		if len(val) == 0 {
+			err = st.Delete(key.Bytes())
+		} else {
+			err = st.Update(key.Bytes(), val)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	root, result, err := st.Commit(false)
+	if err != nil {
+		return err
+	}
+	if root != prev.Root {
+		return errors.New("failed to reset storage trie")
+	}
+	// The returned set can be nil if storage trie is not changed
+	// at all.
+	if result != nil {
+		if err := ctx.nodes.Merge(result); err != nil {
+			return err
+		}
+	}
+	// Write the prev-state account into the main trie
+	full, err := rlp.EncodeToBytes(prev)
+	if err != nil {
+		return err
+	}
+	return ctx.accountTrie.Update(addrHash.Bytes(), full)
+}
+
+// deleteAccount the account was not present in prev-state, and is expected
+// to be existent in post-state. Apply the reverse diff and verify if the
+// account and storage is wiped out correctly.
+func deleteAccount(ctx *context, loader TrieLoader, addr common.Address) error {
+	// The account must be existent in post-state, load the account.
+	h := newHasher()
+	defer h.release()
+
+	addrHash := h.hash(addr.Bytes())
+	blob, err := ctx.accountTrie.Get(addrHash.Bytes())
+	if err != nil {
+		return err
+	}
+	if len(blob) == 0 {
+		return fmt.Errorf("account is non-existent %#x", addrHash)
+	}
+	var post types.StateAccount
+	if err := rlp.DecodeBytes(blob, &post); err != nil {
+		return err
+	}
+	st, err := loader.OpenStorageTrie(ctx.postRoot, addrHash, post.Root)
+	if err != nil {
+		return err
+	}
+	for key, val := range ctx.storages[addr] {
+		if len(val) != 0 {
+			return errors.New("expect storage deletion")
+		}
+		if err := st.Delete(key.Bytes()); err != nil {
+			return err
+		}
+	}
+	root, result, err := st.Commit(false)
+	if err != nil {
+		return err
+	}
+	if root != types.EmptyRootHash {
+		return errors.New("failed to clear storage trie")
+	}
+	// The returned set can be nil if storage trie is not changed
+	// at all.
+	if result != nil {
+		if err := ctx.nodes.Merge(result); err != nil {
+			return err
+		}
+	}
+	// Delete the post-state account from the main trie.
+	return ctx.accountTrie.Delete(addrHash.Bytes())
+}
+
+// hasher is used to compute the sha256 hash of the provided data.
+type hasher struct{ sha crypto.KeccakState }
+
+var hasherPool = sync.Pool{
+	New: func() interface{} { return &hasher{sha: crypto.NewKeccakState()} },
+}
+
+func newHasher() *hasher {
+	return hasherPool.Get().(*hasher)
+}
+
+func (h *hasher) hash(data []byte) common.Hash {
+	return crypto.HashData(h.sha, data)
+}
+
+func (h *hasher) release() {
+	hasherPool.Put(h)
+}

```